### PR TITLE
add feature of different number of points per element #37

### PR DESCRIPTION
**Pull request by ezrayst**
Issue #37

### CODE REVIEW

**Comment by ezrayst** [comment]
Still to do:
(1) `compute_stress` is something weird with `max_height` in range-based loop
(2) io for `ngauss_points` need correction
(3) tidy up the element `hexahedron`

File: include/gmsh.h [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
+#include <array>
+#include <algorithm>
 #include <fstream>
 #include <iostream>
 #include <limits>
 #include <map>
 #include <memory>
 #include <utility>
 #include <vector>
+#include <cmath>
```
Comment (kks32) [inline include/gmsh.h]: `cmath` should be at the top see: https://cpp.cb-geo.com/header/order_includes.html

File: include/gmsh.h [hunk 2]
```diff
@@ -35,17 +40,24 @@ class GMSH : public Mesh<Tdim, Tvertices> {
   void store_element_vertices();
 
   //! Compute material points from element coordinate map
-  void compute_material_points();
+  void compute_material_points(const unsigned ngauss_points);
```
Comment (kks32) [inline include/gmsh.h]: `unsigned` pass by value, no const necessary.

File: include/gmsh.tcc [hunk 3]
```diff
@@ -197,49 +197,90 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::compute_material_points() {
+void GMSH<Tdim, Tvertices>::compute_material_points(const unsigned ngauss_points) {
```
Comment (kks32) [inline include/gmsh.tcc]: Remove const.

File: include/gmsh.tcc [hunk 4]
```diff
@@ -197,49 +197,90 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::compute_material_points() {
```
Comment (kks32) [inline include/gmsh.tcc]: Add doxygent comment for `\param[in]`

File: include/hexahedron_element.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#ifndef MPM_HEXAHEDRON_ELEMENT_H
```
Comment (kks32) [inline include/hexahedron_element.h]: Move this to a `namespace element::hexhedron`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.h (hunk 1 above)
Comment (ezrayst) [reply include/gmsh.h]: forgot to clang format...

File: include/gmsh.h (hunk 2 above)
Comment (ezrayst) [reply include/gmsh.h]: why not pass by value for an `unsigned`?

File: include/gmsh.tcc [hunk 6]
```diff
@@ -196,50 +196,89 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+      // Compute gauss point in cartesian coordinate
+      for (unsigned i = 0; i < Tdim; ++i) {
+        pointsarray[i] = 0;
+        Eigen::VectorXd shape_function = element::hexahedron::shapefn(xi);
+        for (unsigned j = 0; j < Tvertices; ++j) {
+          pointsarray[i] += shape_function[j] * node_coordinates(i, j);
+        }
+      }
+
+      // Unused
```
Comment (kks32) [inline include/gmsh.tcc]: Remove this.

File: include/gauss_points.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
+#ifndef MPM_GAUSS_POINTS_H_
+#define MPM_GAUSS_POINTS_H_
+
+#include <map>
+#include <vector>
+
+namespace gauss_points {
+
+//! Map to store gauss points coefficients
+const std::map<unsigned, std::vector<double>> gauss_points{
```
Comment (kks32) [inline include/gauss_points.h]: Move this to `element` namespace

File: include/element.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
+#ifndef MPM_ELEMENT_H
+#define MPM_ELEMENT_H
+
+#include <eigen3/Eigen/Dense>
+
+namespace element {
+
+namespace hexahedron {
+
+Eigen::VectorXd shapefn(const std::array<double, 3>& xi) {
```
Comment (kks32) [inline include/element.h]: `inline` this function.

File: include/gmsh.h [hunk 9]
```diff
@@ -1,6 +1,9 @@
 #ifndef MPM_POINT_GEN_GMSH_H_
 #define MPM_POINT_GEN_GMSH_H_
 
+#include <algorithm>
+#include <array>
+#include <cmath>
```
Comment (kks32) [inline include/gmsh.h]: Move `cmath`

File: include/gmsh.h [hunk 10]
```diff
@@ -35,17 +40,24 @@ class GMSH : public Mesh<Tdim, Tvertices> {
...
+  //! Number of gauss points per coordinate
+  using Mesh<Tdim, Tvertices>::ngauss_points_;
+
   //! Map of vertex id to its coordinates
   using Mesh<Tdim, Tvertices>::vertices_;
+
   //! Map of elemnt id to its vertices
   using Mesh<Tdim, Tvertices>::elements_;
+
   //! Map of element ids to vertices coordinates
```
Comment (kks32) [inline include/gmsh.h]: element coordinates

File: include/material_points.tcc [hunk 11]
```diff
@@ -36,14 +36,15 @@ void MaterialPoints<Tdim>::compute_stress() {
...
+  //! Compute maximum height of the points
   //! [2D], y is the vertical direction
   //! [3D], z is the vertical direction
   //! In general, [Tdim - 1]
+  double max_height = std::numeric_limits<double>::min();
   for (const auto& point : points_) {
-    if (point->coordinates()[Tdim - 1] > max_height) {
-      max_height = point->coordinates()[Tdim - 1];
+    const double height = point->coordinates()[Tdim - 1];
+    if (height > max_height) {
```
Comment (kks32) [inline include/material_points.tcc]: Remove `{` `}`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/element.h [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+     {-0.8611363115940526, -0.3399810435848563, 0.3399810435848563,
+      0.8611363115940526}},
+    {5,
+     {-0.9061798459386640, -0.5384693101056831, 0, 0.5384693101056831,
+      0.9061798459386640}}};
+}
+
+namespace hexahedron {
+
+inline Eigen::VectorXd shapefn(const std::array<double, 3>& xi) {
```
Comment (kks32) [inline include/element.h]: Needs doxygen comments here.

File: include/element.h [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    {3, {-0.7745966692414834, 0, 0.7745966692414834}},
+    {4,
+     {-0.8611363115940526, -0.3399810435848563, 0.3399810435848563,
+      0.8611363115940526}},
+    {5,
+     {-0.9061798459386640, -0.5384693101056831, 0, 0.5384693101056831,
+      0.9061798459386640}}};
+}
+
+namespace hexahedron {
```
Comment (kks32) [inline include/element.h]: Describe namespace

File: include/element.h [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+#define MPM_ELEMENT_H
+
+#include <map>
+#include <vector>
+
+#include <eigen3/Eigen/Dense>
+
+namespace element {
+
+namespace gauss_points {
```
Comment (kks32) [inline include/element.h]: No namespace is required here. Just element namespace is sufficient.

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

**Pull request merged**